File: VB/WinWebSolution.Module.Web/HighlightFocusedLayoutItemDetailViewController.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports DevExpress.Web.ASPxClasses
Imports DevExpress.ExpressApp.Editors
Imports DevExpress.ExpressApp.Web.Editors
Imports DevExpress.ExpressApp.Web.Editors.ASPx

Namespace WinWebSolution.Module.Web
	Public Class HighlightFocusedLayoutItemDetailViewController
		Inherits HighlightFocusedLayoutItemDetailViewControllerBase
		Private Const ClientSideEventHandlerFunctionFormat As String = "function(s,e){{{0}}}"
		Protected Overrides Sub OnActivated()
			MyBase.OnActivated()
			If View.ViewEditMode = ViewEditMode.Edit Then
				For Each item As WebPropertyEditor In View.GetItems(Of WebPropertyEditor)()
					If item.Editor IsNot Nothing Then
						ApplyFocusedStyle(item)
					Else
						AddHandler item.ControlCreated, Function(s, e) AnonymousMethod1(s, e)
					End If
				Next item
			End If
		End Sub
		
		Private Function AnonymousMethod1(ByVal s As Object, ByVal e As Object) As Boolean
			ApplyFocusedStyle(s)
			Return True
		End Function
		Protected Overrides Sub ApplyFocusedStyle(ByVal element As Object)
			If TypeOf element Is ASPxLookupPropertyEditor Then
				ApplyFocusedStyleCore((CType(element, ASPxLookupPropertyEditor)).DropDownEdit.DropDown)
				ApplyFocusedStyleCore((CType(element, ASPxLookupPropertyEditor)).FindEdit.TextBox)
			ElseIf TypeOf element Is WebPropertyEditor Then
				ApplyFocusedStyleCore(TryCast((CType(element, WebPropertyEditor)).Editor, ASPxWebControl))
			End If
		End Sub
		Private Sub ApplyFocusedStyleCore(ByVal dxControl As ASPxWebControl)
			If dxControl IsNot Nothing Then
				Dim loadEventHandler As EventHandler = Function(s, e) AnonymousMethod2(s, e)
				Dim disposedEventHandler As EventHandler = Nothing
				disposedEventHandler = Function(s, e) AnonymousMethod3(s, e, disposedEventHandler, loadEventHandler)
				AddHandler dxControl.Disposed, disposedEventHandler
				AddHandler dxControl.Load, loadEventHandler
			End If
		End Sub
		
		Private Function AnonymousMethod2(ByVal s As Object, ByVal e As Object) As Boolean
			Dim control As ASPxWebControl = CType(s, ASPxWebControl)
			AddEventHandlerSafe(control, "Init", "window.initEditor(s,e);")
			AddEventHandlerSafe(control, "GotFocus", "window.gotFocusEditor(s,e);")
			Return True
		End Function
		
		Private Function AnonymousMethod3(ByVal s As Object, ByVal e As Object, ByVal disposedEventHandler As EventHandler, ByVal loadEventHandler As EventHandler) As EventHandler
			Dim control As ASPxWebControl = CType(s, ASPxWebControl)
			RemoveHandler control.Disposed, disposedEventHandler
			RemoveHandler control.Load, loadEventHandler
		End Function
		Private Shared Sub AddEventHandlerSafe(ByVal control As ASPxWebControl, ByVal eventName As String, ByVal handler As String)
			Dim existingHandler As String = control.GetClientSideEventHandler(eventName)
			If String.IsNullOrEmpty(existingHandler) Then
				control.SetClientSideEventHandler(eventName, String.Format(ClientSideEventHandlerFunctionFormat, handler))
			Else
				existingHandler = String.Format("{0}{1}" & Constants.vbCrLf & "}}", existingHandler.Substring(0, existingHandler.LastIndexOf("}"c)), handler)
				control.SetClientSideEventHandler(eventName, existingHandler)
			End If
		End Sub
	End Class
End Namespace

File: CS/WinWebSolution.Module.Web/HighlightFocusedLayoutItemDetailViewController.cs
using System;
using System.Web.UI;
using System.Collections.Generic;
using DevExpress.Web.ASPxClasses;
using DevExpress.ExpressApp.Editors;
using DevExpress.ExpressApp.Web;

namespace WinWebSolution.Module.Web {
    public class HighlightFocusedLayoutItemDetailViewController : HighlightFocusedLayoutItemDetailViewControllerBase {
        private const string ScriptKey = "assignStyleScriptKey";
        const string ScriptFunction = @"
                    window.lastFocusedElement = undefined;
                    window.isLayoutItemRow = function(obj) {
                        return obj.className == ""Item"" && obj.tagName == ""DIV"";
                    };
                    window.isValid = function (obj) {
                        return obj != null && obj != undefined; 
                    };
                    window.assignStyle = function (element, color){
                        var curr = element;
                        while (isValid(curr) && !isLayoutItemRow(curr)){
                            curr = curr.parentNode;
                        };
                        if (isValid(curr)) 
                            curr.style.backgroundColor = color;
                    };";
        protected override void OnActivated() {
            base.OnActivated();
            if (View.ViewEditMode == ViewEditMode.Edit)
                foreach (ViewItem item in View.Items) {
                    item.ControlCreated += delegate(object sender, EventArgs e) {
                        AssignStyle(((ViewItem)sender).Control);
                    };
                }
            WebWindow.CurrentRequestWindow.RegisterClientScript(ScriptKey, ScriptFunction);
        }
        protected override void AssignStyle(object control) {
            foreach (ASPxWebControl dxEditor in FindNestedControls<ASPxWebControl>(control as Control)) {
                string functionInit = @"
                function (s, e){
                    if (s.focused) {
                        window.assignStyle(window.lastFocusedElement = s.GetMainElement(), ""#C9E4F0"");
                    }
                }";
                string functionGotFocus = @"
                function (s, e){
                    window.assignStyle(window.lastFocusedElement, """");
                    window.assignStyle(window.lastFocusedElement = s.GetMainElement(), ""#C9E4F0"");
                }";
                if (string.IsNullOrEmpty(dxEditor.GetClientSideEventHandler("Init")))
                    dxEditor.SetClientSideEventHandler("Init", functionInit);
                if (string.IsNullOrEmpty(dxEditor.GetClientSideEventHandler("GotFocus")))
                    dxEditor.SetClientSideEventHandler("GotFocus", functionGotFocus);
            }
        }
        private static IEnumerable<TNestedControl> FindNestedControls<TNestedControl>(Control container) where TNestedControl : Control {
            if (container != null && container.Controls != null)
                foreach (Control item in container.Controls) {
                    if (item is TNestedControl)
                        yield return (TNestedControl)item;
                    foreach (TNestedControl child in FindNestedControls<TNestedControl>(item))
                        yield return child;
                }
        }
    }
}